File: storage.py
from __future__ import annotations

import csv
import json
import sqlite3
from datetime import datetime, timezone
from io import StringIO
from pathlib import Path
from typing import Any

from csi import FACTORS, ITEMS, PAIR_COMPARISONS


DB_PATH = Path("data/csi.sqlite3")
EXPERIMENT_CONDITION_NAMES = (
    "統制条件",
    "実験条件１（自律）",
    "実験条件２（他律）",
)


def connect(db_path: Path = DB_PATH) -> sqlite3.Connection:
    db_path.parent.mkdir(parents=True, exist_ok=True)
    conn = sqlite3.connect(db_path, check_same_thread=False)
    conn.row_factory = sqlite3.Row
    return conn


def init_db(conn: sqlite3.Connection) -> None:
    conn.executescript(
        """
        CREATE TABLE IF NOT EXISTS conditions (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            name TEXT NOT NULL UNIQUE,
            created_at TEXT NOT NULL
        );
        """
    )
    reset_legacy_response_schema(conn)
    conn.executescript(
        """

        CREATE TABLE IF NOT EXISTS responses (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            participant_id TEXT NOT NULL,
            condition_id INTEGER NOT NULL,
            submitted_at TEXT NOT NULL,
            score_type TEXT NOT NULL,
            item_scores_json TEXT NOT NULL,
            pair_choices_json TEXT NOT NULL,
            pair_order_json TEXT,
            factor_scores_json TEXT NOT NULL,
            factor_counts_json TEXT NOT NULL,
            weighted_scores_json TEXT NOT NULL,
            csi_score REAL NOT NULL,
            FOREIGN KEY (condition_id) REFERENCES conditions(id)
        );
        """
    )
    ensure_response_schema(conn)
    ensure_experiment_conditions(conn)
    conn.commit()


def response_schema_columns(conn: sqlite3.Connection) -> set[str]:
    return {
        row["name"]
        for row in conn.execute("PRAGMA table_info(responses)").fetchall()
    }


def reset_legacy_response_schema(conn: sqlite3.Connection) -> None:
    columns = response_schema_columns(conn)
    if "collaboration_status" in columns:
        conn.execute("DROP TABLE responses")
        conn.commit()


def ensure_response_schema(conn: sqlite3.Connection) -> None:
    columns = response_schema_columns(conn)
    if "pair_order_json" not in columns:
        conn.execute("ALTER TABLE responses ADD COLUMN pair_order_json TEXT")
        conn.commit()


def utc_now() -> str:
    return datetime.now(timezone.utc).isoformat(timespec="seconds")


def ensure_experiment_conditions(conn: sqlite3.Connection) -> None:
    for name in EXPERIMENT_CONDITION_NAMES:
        conn.execute(
            "INSERT OR IGNORE INTO conditions (name, created_at) VALUES (?, ?)",
            (name, utc_now()),
        )
    conn.commit()


def add_condition(conn: sqlite3.Connection, name: str) -> None:
    cleaned = name.strip()
    if not cleaned:
        raise ValueError("条件名を入力してください。")
    conn.execute(
        "INSERT INTO conditions (name, created_at) VALUES (?, ?)",
        (cleaned, utc_now()),
    )
    conn.commit()


def list_conditions(conn: sqlite3.Connection) -> list[sqlite3.Row]:
    return conn.execute(
        "SELECT id, name, created_at FROM conditions ORDER BY id"
    ).fetchall()


def list_experiment_conditions(conn: sqlite3.Connection) -> list[sqlite3.Row]:
    ensure_experiment_conditions(conn)
    placeholders = ",".join("?" for _ in EXPERIMENT_CONDITION_NAMES)
    return conn.execute(
        f"""
        SELECT id, name, created_at
        FROM conditions
        WHERE name IN ({placeholders})
        ORDER BY CASE name
            WHEN ? THEN 1
            WHEN ? THEN 2
            WHEN ? THEN 3
            ELSE 4
        END
        """,
        (*EXPERIMENT_CONDITION_NAMES, *EXPERIMENT_CONDITION_NAMES),
    ).fetchall()


def condition_response_count(conn: sqlite3.Connection, condition_id: int) -> int:
    row = conn.execute(
        "SELECT COUNT(*) AS response_count FROM responses WHERE condition_id = ?",
        (condition_id,),
    ).fetchone()
    return int(row["response_count"])


def delete_condition(conn: sqlite3.Connection, condition_id: int) -> int:
    if condition_response_count(conn, condition_id) > 0:
        raise ValueError("この条件には回答が紐づいているため削除できません。先に該当する回答を削除してください。")
    cur = conn.execute("DELETE FROM conditions WHERE id = ?", (condition_id,))
    conn.commit()
    return int(cur.rowcount)


def delete_response(conn: sqlite3.Connection, response_id: int) -> int:
    cur = conn.execute("DELETE FROM responses WHERE id = ?", (response_id,))
    conn.commit()
    return int(cur.rowcount)


def save_response(
    conn: sqlite3.Connection,
    participant_id: str,
    condition_id: int,
    item_scores: dict[str, int],
    pair_choices: list[str],
    scores: dict[str, Any],
    pair_order: list[int] | None = None,
) -> int:
    ensure_response_schema(conn)
    if pair_order is None:
        pair_order = list(range(len(pair_choices)))
    cur = conn.execute(
        """
        INSERT INTO responses (
            participant_id,
            condition_id,
            submitted_at,
            score_type,
            item_scores_json,
            pair_choices_json,
            pair_order_json,
            factor_scores_json,
            factor_counts_json,
            weighted_scores_json,
            csi_score
        ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        """,
        (
            participant_id,
            condition_id,
            utc_now(),
            scores["score_type"],
            json.dumps(item_scores, ensure_ascii=False),
            json.dumps(pair_choices, ensure_ascii=False),
            json.dumps(pair_order, ensure_ascii=False),
            json.dumps(scores["factor_scores"], ensure_ascii=False),
            json.dumps(scores["factor_counts"], ensure_ascii=False),
            json.dumps(scores["weighted_scores"], ensure_ascii=False),
            scores["csi_score"],
        ),
    )
    conn.commit()
    return int(cur.lastrowid)


def response_rows(conn: sqlite3.Connection) -> list[sqlite3.Row]:
    return conn.execute(
        """
        SELECT
            r.id,
            r.participant_id,
            r.condition_id,
            c.name AS condition_name,
            r.submitted_at,
            r.score_type,
            r.item_scores_json,
            r.pair_choices_json,
            r.pair_order_json,
            r.factor_scores_json,
            r.factor_counts_json,
            r.weighted_scores_json,
            r.csi_score
        FROM responses r
        JOIN conditions c ON c.id = r.condition_id
        ORDER BY r.id
        """
    ).fetchall()


def condition_summary(conn: sqlite3.Connection) -> list[sqlite3.Row]:
    placeholders = ",".join("?" for _ in EXPERIMENT_CONDITION_NAMES)
    return conn.execute(
        f"""
        SELECT
            c.id AS condition_id,
            c.name AS condition_name,
            COUNT(r.id) AS response_count,
            AVG(r.csi_score) AS csi_mean,
            CASE
                WHEN COUNT(r.id) > 1 THEN
                    sqrt((SUM(r.csi_score * r.csi_score) - SUM(r.csi_score) * SUM(r.csi_score) / COUNT(r.id)) / (COUNT(r.id) - 1))
                ELSE 0
            END AS csi_sd
        FROM conditions c
        LEFT JOIN responses r ON r.condition_id = c.id
        WHERE c.name IN ({placeholders})
        GROUP BY c.id, c.name
        ORDER BY CASE c.name
            WHEN ? THEN 1
            WHEN ? THEN 2
            WHEN ? THEN 3
            ELSE 4
        END
        """,
        (*EXPERIMENT_CONDITION_NAMES, *EXPERIMENT_CONDITION_NAMES),
    ).fetchall()


def participant_completion_summary(conn: sqlite3.Connection) -> list[dict[str, Any]]:
    rows = response_rows(conn)
    fixed_names = set(EXPERIMENT_CONDITION_NAMES)
    grouped: dict[str, set[str]] = {}
    for row in rows:
        if row["condition_name"] in fixed_names:
            grouped.setdefault(row["participant_id"], set()).add(row["condition_name"])

    summary = []
    for participant_id, completed_names in sorted(grouped.items()):
        summary.append(
            {
                "participant_id": participant_id,
                "completed_conditions": len(completed_names),
                "total_conditions": len(EXPERIMENT_CONDITION_NAMES),
                "is_complete": len(completed_names) == len(EXPERIMENT_CONDITION_NAMES),
                "missing_conditions": ", ".join(
                    name
                    for name in EXPERIMENT_CONDITION_NAMES
                    if name not in completed_names
                ),
            }
        )
    return summary


def factor_summary(conn: sqlite3.Connection) -> list[dict[str, Any]]:
    rows = response_rows(conn)
    grouped: dict[tuple[int, str, str], list[float]] = {}
    for row in rows:
        factor_scores = json.loads(row["factor_scores_json"])
        for factor in FACTORS:
            key = (row["condition_id"], row["condition_name"], factor)
            grouped.setdefault(key, []).append(float(factor_scores.get(factor, 0)))

    summary = []
    for (condition_id, condition_name, factor), values in sorted(grouped.items()):
        mean = sum(values) / len(values)
        if len(values) > 1:
            variance = sum((value - mean) ** 2 for value in values) / (len(values) - 1)
            sd = variance ** 0.5
        else:
            sd = 0.0
        summary.append(
            {
                "condition_id": condition_id,
                "condition_name": condition_name,
                "factor": factor,
                "response_count": len(values),
                "factor_score_mean": mean,
                "factor_score_sd": sd,
            }
        )
    return summary


def csv_text(rows: list[dict[str, Any]]) -> str:
    if not rows:
        return ""
    output = StringIO()
    writer = csv.DictWriter(output, fieldnames=list(rows[0].keys()), lineterminator="\n")
    writer.writeheader()
    writer.writerows(rows)
    return output.getvalue()


def raw_data_csv(conn: sqlite3.Connection) -> str:
    rows = []
    for row in response_rows(conn):
        item_scores = json.loads(row["item_scores_json"])
        factor_scores = json.loads(row["factor_scores_json"])
        factor_counts = json.loads(row["factor_counts_json"])
        weighted_scores = json.loads(row["weighted_scores_json"])
        flat = {
            "response_id": row["id"],
            "participant_id": row["participant_id"],
            "condition_id": row["condition_id"],
            "condition_name": row["condition_name"],
            "submitted_at": row["submitted_at"],
            "score_type": row["score_type"],
            "csi_score": row["csi_score"],
        }
        for item in ITEMS:
            flat[item["id"]] = item_scores.get(item["id"]) if item["scoreable"] else "N/A"
        for factor in FACTORS:
            flat[f"{factor}_score"] = factor_scores.get(factor, 0)
            flat[f"{factor}_count"] = factor_counts.get(factor, 0)
            flat[f"{factor}_weighted"] = weighted_scores.get(factor, 0)
        rows.append(flat)
    return csv_text(rows)


def item_data_csv(conn: sqlite3.Connection) -> str:
    rows = []
    for row in response_rows(conn):
        item_scores = json.loads(row["item_scores_json"])
        for item in ITEMS:
            rows.append(
                {
                    "response_id": row["id"],
                    "participant_id": row["participant_id"],
                    "condition_id": row["condition_id"],
                    "condition_name": row["condition_name"],
                    "submitted_at": row["submitted_at"],
                    "item_id": item["id"],
                    "factor": item["factor"],
                    "item_text_ja": item["text_ja"],
                    "score": item_scores[item["id"]] if item["scoreable"] else "N/A",
                }
            )
    return csv_text(rows)


def factor_data_csv(conn: sqlite3.Connection) -> str:
    rows = []
    for row in response_rows(conn):
        factor_scores = json.loads(row["factor_scores_json"])
        factor_counts = json.loads(row["factor_counts_json"])
        weighted_scores = json.loads(row["weighted_scores_json"])
        for factor in FACTORS:
            rows.append(
                {
                    "response_id": row["id"],
                    "participant_id": row["participant_id"],
                    "condition_id": row["condition_id"],
                    "condition_name": row["condition_name"],
                    "submitted_at": row["submitted_at"],
                    "score_type": row["score_type"],
                    "factor": factor,
                    "factor_score": factor_scores.get(factor, 0),
                    "factor_count": factor_counts.get(factor, 0),
                    "weighted_factor_score": weighted_scores.get(factor, 0),
                    "csi_score": row["csi_score"],
                }
            )
    return csv_text(rows)


def condition_summary_csv(conn: sqlite3.Connection) -> str:
    rows = [dict(row) for row in condition_summary(conn)]
    return csv_text(rows)


def pair_data_csv(conn: sqlite3.Connection) -> str:
    rows = []
    for row in response_rows(conn):
        pair_choices = json.loads(row["pair_choices_json"])
        raw_pair_order = row["pair_order_json"] if "pair_order_json" in row.keys() else None
        pair_order = json.loads(raw_pair_order) if raw_pair_order else list(range(len(pair_choices)))
        for index, choice in enumerate(pair_choices):
            original_index = pair_order[index] if index < len(pair_order) else index
            pair = (
                PAIR_COMPARISONS[original_index]
                if 0 <= original_index < len(PAIR_COMPARISONS)
                else ("", "")
            )
            rows.append(
                {
                    "response_id": row["id"],
                    "participant_id": row["participant_id"],
                    "condition_id": row["condition_id"],
                    "condition_name": row["condition_name"],
                    "submitted_at": row["submitted_at"],
                    "pair_index": index + 1,
                    "original_pair_index": original_index + 1 if pair[0] else "",
                    "factor_a": pair[0],
                    "factor_b": pair[1],
                    "chosen_factor": choice,
                }
            )
    return csv_text(rows)

File: lib/storage.ts
import { mkdirSync } from "node:fs";
import path from "node:path";
import { DatabaseSync } from "node:sqlite";
import { calculateScores, FACTORS, Factor, ITEMS, PAIR_COMPARISONS, CsiScores } from "@/lib/csi";

export const DB_PATH = path.join(process.cwd(), "data", "csi.sqlite3");
export const EXPERIMENT_CONDITION_NAMES = [
  "条件１：なるべく身体を動かさない",
  "条件２：貧乏ゆすりをしながら",
  "条件３：エアマットで揺らされながら",
] as const;

export type ConditionRow = {
  id: number;
  name: string;
  created_at: string;
};

export type ResponseRow = {
  id: number;
  participant_id: string;
  condition_id: number;
  condition_name: string;
  submitted_at: string;
  score_type: string;
  item_scores_json: string;
  pair_choices_json: string;
  pair_order_json: string | null;
  factor_scores_json: string;
  factor_counts_json: string;
  weighted_scores_json: string;
  csi_score: number;
};

export type ParticipantResumeSummary = {
  participant_id: string;
  completed_conditions: number;
  total_conditions: number;
  missing_conditions: string[];
};

type DatabaseConnection = DatabaseSync;

let db: DatabaseConnection | null = null;

export function utcNow(): string {
  return new Date().toISOString().replace(/\.\d{3}Z$/, "+00:00");
}

export function connect(dbPath = DB_PATH): DatabaseConnection {
  if (dbPath === DB_PATH && db) {
    return db;
  }
  mkdirSync(path.dirname(dbPath), { recursive: true });
  const conn = new DatabaseSync(dbPath);
  conn.exec("PRAGMA foreign_keys = ON");
  initDb(conn);
  if (dbPath === DB_PATH) {
    db = conn;
  }
  return conn;
}

export function initDb(conn: DatabaseConnection): void {
  conn.exec(`
    CREATE TABLE IF NOT EXISTS conditions (
      id INTEGER PRIMARY KEY AUTOINCREMENT,
      name TEXT NOT NULL UNIQUE,
      created_at TEXT NOT NULL
    );
  `);
  resetLegacyResponseSchema(conn);
  conn.exec(`
    CREATE TABLE IF NOT EXISTS responses (
      id INTEGER PRIMARY KEY AUTOINCREMENT,
      participant_id TEXT NOT NULL,
      condition_id INTEGER NOT NULL,
      submitted_at TEXT NOT NULL,
      score_type TEXT NOT NULL,
      item_scores_json TEXT NOT NULL,
      pair_choices_json TEXT NOT NULL,
      pair_order_json TEXT,
      factor_scores_json TEXT NOT NULL,
      factor_counts_json TEXT NOT NULL,
      weighted_scores_json TEXT NOT NULL,
      csi_score REAL NOT NULL,
      FOREIGN KEY (condition_id) REFERENCES conditions(id)
    );
  `);
  ensureResponseSchema(conn);
  ensureExperimentConditions(conn);
}

export function responseSchemaColumns(conn: DatabaseConnection): Set<string> {
  const rows = conn.prepare("PRAGMA table_info(responses)").all() as { name: string }[];
  return new Set(rows.map((row) => row.name));
}

export function resetLegacyResponseSchema(conn: DatabaseConnection): void {
  const columns = responseSchemaColumns(conn);
  if (columns.has("collaboration_status")) {
    conn.exec("DROP TABLE responses");
  }
}

export function ensureResponseSchema(conn: DatabaseConnection): void {
  const columns = responseSchemaColumns(conn);
  if (columns.size > 0 && !columns.has("pair_order_json")) {
    conn.exec("ALTER TABLE responses ADD COLUMN pair_order_json TEXT");
  }
}

export function ensureExperimentConditions(conn: DatabaseConnection): void {
  const stmt = conn.prepare("INSERT OR IGNORE INTO conditions (name, created_at) VALUES (?, ?)");
  conn.exec("BEGIN");
  try {
    for (const name of EXPERIMENT_CONDITION_NAMES) {
      stmt.run(name, utcNow());
    }
    conn.exec("COMMIT");
  } catch (error) {
    conn.exec("ROLLBACK");
    throw error;
  }
}

export function listExperimentConditions(conn = connect()): ConditionRow[] {
  ensureExperimentConditions(conn);
  return conn
    .prepare(
      `
      SELECT id, name, created_at
      FROM conditions
      WHERE name IN (${EXPERIMENT_CONDITION_NAMES.map(() => "?").join(",")})
      ORDER BY CASE name
        WHEN ? THEN 1
        WHEN ? THEN 2
        WHEN ? THEN 3
        ELSE 4
      END
      `,
    )
    .all(...EXPERIMENT_CONDITION_NAMES, ...EXPERIMENT_CONDITION_NAMES) as ConditionRow[];
}

export function conditionResponseCount(conn: DatabaseConnection, conditionId: number): number {
  const row = conn
    .prepare("SELECT COUNT(*) AS response_count FROM responses WHERE condition_id = ?")
    .get(conditionId) as { response_count: number };
  return Number(row.response_count);
}

export function deleteCondition(conn: DatabaseConnection, conditionId: number): number {
  if (conditionResponseCount(conn, conditionId) > 0) {
    throw new Error("この条件には回答が紐づいているため削除できません。先に該当する回答を削除してください。");
  }
  return Number(conn.prepare("DELETE FROM conditions WHERE id = ?").run(conditionId).changes);
}

export function deleteResponse(conn: DatabaseConnection, responseId: number): number {
  return Number(conn.prepare("DELETE FROM responses WHERE id = ?").run(responseId).changes);
}

export function deleteParticipantResponses(conn: DatabaseConnection, participantId: string): number {
  return Number(conn.prepare("DELETE FROM responses WHERE participant_id = ?").run(participantId).changes);
}

export function mergeParticipantResponses(
  conn: DatabaseConnection,
  sourceParticipantId: string,
  targetParticipantId: string,
): number {
  return Number(
    conn
      .prepare("UPDATE responses SET participant_id = ? WHERE participant_id = ?")
      .run(targetParticipantId, sourceParticipantId).changes,
  );
}

export function saveResponse(
  conn: DatabaseConnection,
  participantId: string,
  conditionId: number,
  itemScores: Record<string, number>,
  pairChoices: Factor[],
  scores: CsiScores = calculateScores(itemScores, pairChoices),
  pairOrder: number[] | null = null,
): number {
  ensureResponseSchema(conn);
  const normalizedPairOrder = pairOrder ?? Array.from({ length: pairChoices.length }, (_, index) => index);
  const info = conn
    .prepare(
      `
      INSERT INTO responses (
        participant_id,
        condition_id,
        submitted_at,
        score_type,
        item_scores_json,
        pair_choices_json,
        pair_order_json,
        factor_scores_json,
        factor_counts_json,
        weighted_scores_json,
        csi_score
      ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
      `,
    )
    .run(
      participantId,
      conditionId,
      utcNow(),
      scores.scoreType,
      JSON.stringify(itemScores),
      JSON.stringify(pairChoices),
      JSON.stringify(normalizedPairOrder),
      JSON.stringify(scores.factorScores),
      JSON.stringify(scores.factorCounts),
      JSON.stringify(scores.weightedScores),
      scores.csiScore,
    );
  return Number(info.lastInsertRowid);
}

export function responseRows(conn = connect()): ResponseRow[] {
  return conn
    .prepare(
      `
      SELECT
        r.id,
        r.participant_id,
        r.condition_id,
        c.name AS condition_name,
        r.submitted_at,
        r.score_type,
        r.item_scores_json,
        r.pair_choices_json,
        r.pair_order_json,
        r.factor_scores_json,
        r.factor_counts_json,
        r.weighted_scores_json,
        r.csi_score
      FROM responses r
      JOIN conditions c ON c.id = r.condition_id
      ORDER BY r.id
      `,
    )
    .all() as ResponseRow[];
}

export function completedConditionIdsForParticipant(conn: DatabaseConnection, participantId: string): number[] {
  const conditions = listExperimentConditions(conn);
  const conditionIds = new Set(conditions.map((condition) => condition.id));
  const rows = conn
    .prepare(
      `
      SELECT DISTINCT condition_id
      FROM responses
      WHERE participant_id = ?
      ORDER BY condition_id
      `,
    )
    .all(participantId) as { condition_id: number }[];
  return rows.map((row) => row.condition_id).filter((conditionId) => conditionIds.has(conditionId));
}

export function incompleteParticipantSummaries(conn = connect()): ParticipantResumeSummary[] {
  const conditions = listExperimentConditions(conn);
  const conditionIds = new Set(conditions.map((condition) => condition.id));
  const grouped = new Map<string, Set<number>>();
  for (const row of responseRows(conn)) {
    if (!conditionIds.has(row.condition_id)) {
      continue;
    }
    const completed = grouped.get(row.participant_id) ?? new Set<number>();
    completed.add(row.condition_id);
    grouped.set(row.participant_id, completed);
  }
  return [...grouped.entries()]
    .filter(([, completedIds]) => completedIds.size < conditions.length)
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([participantId, completedIds]) => ({
      participant_id: participantId,
      completed_conditions: completedIds.size,
      total_conditions: conditions.length,
      missing_conditions: conditions
        .filter((condition) => !completedIds.has(condition.id))
        .map((condition) => condition.name),
    }));
}

export function conditionSummary(conn = connect()): Record<string, unknown>[] {
  return conn
    .prepare(
      `
      SELECT
        c.id AS condition_id,
        c.name AS condition_name,
        COUNT(r.id) AS response_count,
        AVG(r.csi_score) AS csi_mean,
        CASE
          WHEN COUNT(r.id) > 1 THEN
            sqrt((SUM(r.csi_score * r.csi_score) - SUM(r.csi_score) * SUM(r.csi_score) / COUNT(r.id)) / (COUNT(r.id) - 1))
          ELSE 0
        END AS csi_sd
      FROM conditions c
      LEFT JOIN responses r ON r.condition_id = c.id
      WHERE c.name IN (${EXPERIMENT_CONDITION_NAMES.map(() => "?").join(",")})
      GROUP BY c.id, c.name
      ORDER BY CASE c.name
        WHEN ? THEN 1
        WHEN ? THEN 2
        WHEN ? THEN 3
        ELSE 4
      END
      `,
    )
    .all(...EXPERIMENT_CONDITION_NAMES, ...EXPERIMENT_CONDITION_NAMES) as Record<string, unknown>[];
}

export function participantCompletionSummary(conn = connect()): Record<string, unknown>[] {
  const fixedNames = new Set<string>(EXPERIMENT_CONDITION_NAMES);
  const grouped = new Map<string, Set<string>>();
  for (const row of responseRows(conn)) {
    if (fixedNames.has(row.condition_name)) {
      const completed = grouped.get(row.participant_id) ?? new Set<string>();
      completed.add(row.condition_name);
      grouped.set(row.participant_id, completed);
    }
  }
  return [...grouped.entries()]
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([participantId, completedNames]) => ({
      participant_id: participantId,
      completed_conditions: completedNames.size,
      total_conditions: EXPERIMENT_CONDITION_NAMES.length,
      is_complete: completedNames.size === EXPERIMENT_CONDITION_NAMES.length,
      missing_conditions: EXPERIMENT_CONDITION_NAMES.filter((name) => !completedNames.has(name)).join(", "),
    }));
}

export function factorSummary(conn = connect()): Record<string, unknown>[] {
  const grouped = new Map<string, { conditionId: number; conditionName: string; factor: Factor; values: number[] }>();
  for (const row of responseRows(conn)) {
    const factorScores = JSON.parse(row.factor_scores_json) as Partial<Record<Factor, number>>;
    for (const factor of FACTORS) {
      const key = `${row.condition_id}\u0000${row.condition_name}\u0000${factor}`;
      const entry =
        grouped.get(key) ?? { conditionId: row.condition_id, conditionName: row.condition_name, factor, values: [] };
      entry.values.push(Number(factorScores[factor] ?? 0));
      grouped.set(key, entry);
    }
  }
  return [...grouped.values()]
    .sort((a, b) => a.conditionId - b.conditionId || FACTORS.indexOf(a.factor) - FACTORS.indexOf(b.factor))
    .map((entry) => {
      const mean = entry.values.reduce((sum, value) => sum + value, 0) / entry.values.length;
      const sd =
        entry.values.length > 1
          ? Math.sqrt(entry.values.reduce((sum, value) => sum + (value - mean) ** 2, 0) / (entry.values.length - 1))
          : 0;
      return {
        condition_id: entry.conditionId,
        condition_name: entry.conditionName,
        factor: entry.factor,
        response_count: entry.values.length,
        factor_score_mean: mean,
        factor_score_sd: sd,
      };
    });
}

function csvEscape(value: unknown): string {
  const text = value == null ? "" : String(value);
  return /[",\n\r]/.test(text) ? `"${text.replaceAll('"', '""')}"` : text;
}

export function csvText(rows: Record<string, unknown>[]): string {
  if (rows.length === 0) {
    return "\uFEFF";
  }
  const headers = Object.keys(rows[0]);
  const body = [
    headers.map(csvEscape).join(","),
    ...rows.map((row) => headers.map((header) => csvEscape(row[header])).join(",")),
  ].join("\r\n") + "\r\n";
  return `\uFEFF${body}`;
}

export function rawDataCsv(conn = connect()): string {
  const rows = responseRows(conn).map((row) => {
    const itemScores = JSON.parse(row.item_scores_json) as Record<string, number>;
    const factorScores = JSON.parse(row.factor_scores_json) as Partial<Record<Factor, number>>;
    const factorCounts = JSON.parse(row.factor_counts_json) as Partial<Record<Factor, number>>;
    const weightedScores = JSON.parse(row.weighted_scores_json) as Partial<Record<Factor, number>>;
    const flat: Record<string, unknown> = {
      response_id: row.id,
      participant_id: row.participant_id,
      condition_id: row.condition_id,
      condition_name: row.condition_name,
      submitted_at: row.submitted_at,
      score_type: row.score_type,
      csi_score: row.csi_score,
    };
    for (const item of ITEMS) {
      flat[item.id] = item.scoreable ? itemScores[item.id] : "N/A";
    }
    for (const factor of FACTORS) {
      flat[`${factor}_score`] = factorScores[factor] ?? 0;
      flat[`${factor}_count`] = factorCounts[factor] ?? 0;
      flat[`${factor}_weighted`] = weightedScores[factor] ?? 0;
    }
    return flat;
  });
  return csvText(rows);
}

export function participantScoresCsv(conn = connect()): string {
  const conditions = listExperimentConditions(conn);
  const latestByParticipant = new Map<string, Map<number, ResponseRow>>();
  for (const row of responseRows(conn)) {
    const participantRows = latestByParticipant.get(row.participant_id) ?? new Map<number, ResponseRow>();
    const current = participantRows.get(row.condition_id);
    if (!current || row.id > current.id) {
      participantRows.set(row.condition_id, row);
    }
    latestByParticipant.set(row.participant_id, participantRows);
  }

  const rows = [...latestByParticipant.entries()]
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([participantId, conditionRows]) => {
      const flat: Record<string, unknown> = { participant_id: participantId };
      conditions.forEach((condition, index) => {
        flat[`condition_${index + 1}_csi_score`] = conditionRows.get(condition.id)?.csi_score ?? "";
      });
      return flat;
    });
  return csvText(rows);
}

export function itemDataCsv(conn = connect()): string {
  const rows = responseRows(conn).flatMap((row) => {
    const itemScores = JSON.parse(row.item_scores_json) as Record<string, number>;
    return ITEMS.map((item) => ({
      response_id: row.id,
      participant_id: row.participant_id,
      condition_id: row.condition_id,
      condition_name: row.condition_name,
      submitted_at: row.submitted_at,
      item_id: item.id,
      factor: item.factor,
      item_text_ja: item.textJa,
      score: item.scoreable ? itemScores[item.id] : "N/A",
    }));
  });
  return csvText(rows);
}

export function factorDataCsv(conn = connect()): string {
  const rows = responseRows(conn).flatMap((row) => {
    const factorScores = JSON.parse(row.factor_scores_json) as Partial<Record<Factor, number>>;
    const factorCounts = JSON.parse(row.factor_counts_json) as Partial<Record<Factor, number>>;
    const weightedScores = JSON.parse(row.weighted_scores_json) as Partial<Record<Factor, number>>;
    return FACTORS.map((factor) => ({
      response_id: row.id,
      participant_id: row.participant_id,
      condition_id: row.condition_id,
      condition_name: row.condition_name,
      submitted_at: row.submitted_at,
      score_type: row.score_type,
      factor,
      factor_score: factorScores[factor] ?? 0,
      factor_count: factorCounts[factor] ?? 0,
      weighted_factor_score: weightedScores[factor] ?? 0,
      csi_score: row.csi_score,
    }));
  });
  return csvText(rows);
}

export function conditionSummaryCsv(conn = connect()): string {
  return csvText(conditionSummary(conn));
}

export function pairDataCsv(conn = connect()): string {
  const rows = responseRows(conn).flatMap((row) => {
    const pairChoices = JSON.parse(row.pair_choices_json) as Factor[];
    const pairOrder = row.pair_order_json
      ? (JSON.parse(row.pair_order_json) as number[])
      : Array.from({ length: pairChoices.length }, (_, index) => index);
    return pairChoices.map((choice, index) => {
      const originalIndex = pairOrder[index] ?? index;
      const pair = originalIndex >= 0 && originalIndex < PAIR_COMPARISONS.length ? PAIR_COMPARISONS[originalIndex] : ["", ""];
      return {
        response_id: row.id,
        participant_id: row.participant_id,
        condition_id: row.condition_id,
        condition_name: row.condition_name,
        submitted_at: row.submitted_at,
        pair_index: index + 1,
        original_pair_index: pair[0] ? originalIndex + 1 : "",
        factor_a: pair[0],
        factor_b: pair[1],
        chosen_factor: choice,
      };
    });
  });
  return csvText(rows);
}
